podvm: Extract podvm smoke test into shell script

ldoktor opened the pull request:
having a dedicated shellscript should allow users (with a big fat warning) to test their images before sending them to cloud. Also it should help maintaining (shellcheck) the code itself.


~Note: Previously my podvms were fine but recently they stopped working. To debug that I used qemu but failed to find the reason. Then I noticed this test so I'd like to make it more visible and ideally add it to README/Makefile so people building their images can have a simple way to test the very basic functionality. Note I don't know why by recently all my attempts are failing as:~

~● attestation-agent.serv… loaded failed     failed       Attestation Agent TTRP…~
~● gen-issue.service       loaded failed     failed       Generate issue to prin…~
~● setup-nat-for-imds.ser… loaded failed     failed       Setup NAT for IMDS~

~And journal complains about missing ns:~

```
Apr 14 13:04:10 fedora systemd[1]: Failed to start agent-protocol-forwarder.service - Agent Protocol Forwarder.
Apr 14 13:04:15 fedora systemd[1]: agent-protocol-forwarder.service: Scheduled restart job, restart counter is at 3.
Apr 14 13:04:15 fedora systemd[1]: Starting agent-protocol-forwarder.service - Agent Protocol Forwarder...
Apr 14 13:04:15 fedora (orwarder)[803]: agent-protocol-forwarder.service: Referenced but unset environment variable evaluates to an empty string: OPTIONS
Apr 14 13:04:15 fedora agent-protocol-forwarder[803]: agent-protocol-forwarder version unknown
Apr 14 13:04:15 fedora agent-protocol-forwarder[803]:   commit: unknown
Apr 14 13:04:15 fedora agent-protocol-forwarder[803]:   go: go1.23.6
Apr 14 13:04:15 fedora agent-protocol-forwarder[803]: 2025/04/14 13:04:15 [podnetwork] Primary interface: ens2, Gateway address: 192.168.124.1
Apr 14 13:04:15 fedora agent-protocol-forwarder[803]: 2025/04/14 13:04:15 [podnetwork] Detected primary interface: ens2
Apr 14 13:04:15 fedora agent-protocol-forwarder[803]: /usr/local/bin/agent-protocol-forwarder: error running a service *forwarder.daemon: failed to set up pod network: failed to open network namespace "/run/netns/podns": failed to get network namespace /run/netns/podns: no such file or directory
Apr 14 13:04:15 fedora systemd[1]: agent-protocol-forwarder.service: Main process exited, code=exited, status=1/FAILURE
Apr 14 13:04:15 fedora systemd[1]: agent-protocol-forwarder.service: Failed with result 'exit-code'.
Apr 14 13:04:15 fedora systemd[1]: Failed to start agent-protocol-forwarder.service - Agent Protocol Forwarder.
```

^^ found out the `make debug` removes the `resources/` dir which removes the `TLS_OPTIONS=-disable-tls` option, causing the podvm image to be not usable.

mkulke commented on src/cloud-api-adaptor/podvm/hack/smoke_test.sh:
```diff
@@ -0,0 +1,179 @@
+# Start the VM
+echo "::debug:: Starting VM"
+# TODO: Add AAVMF for arm
+[ -e "/usr/share/OVMF/OVMF_CODE.fd" ] && OVMF="/usr/share/OVMF/OVMF_CODE.fd"
```
note, this will not work with all OVMF builds, e.g. not with the ones bundled in ubuntu 24.04. the best idea is to download the OVMFs. from a static resource. See the azure podvm-build workflow where we do this for creating the measurements file.

Maybe we don't need to use libvirt for the smoke test? I started with it, because of the libvirt provider, but it's maybe too invasive for local usage, plain qemu could do the same probably.

ldoktor replied:
Well I thought about that too but we'd have to query for IP so I kept using libvirt.

mkulke replied:
what if we use qemu's hostfwd, so the host port can be `HOST_PORT="localhost:15150"`?
```
-netdev type=user,id=user-net,hostfwd=tcp::15150-:15150 \

ldoktor commented on src/cloud-api-adaptor/podvm/hack/smoke_test.sh:
```diff
@@ -0,0 +1,179 @@
+# Requirements listed in ../../../.github/workflows/podvm_smoketest.yaml
+
+DESTRUCTIVE=0
+IMG=""
```
That shouldn't harm

ldoktor commented on src/cloud-api-adaptor/podvm/hack/smoke_test.sh:
```diff
@@ -0,0 +1,179 @@
+	KATA_REF=$(yq -e '.oci.kata-containers.reference' ${SCRIPTDIR}/../../versions.yaml)
+	KATA_REG=$(yq -e '.oci.kata-containers.registry' ${SCRIPTDIR}/../../versions.yaml)
+	echo "::debug:: Pulling kata-ctl from ${KATA_REG}/agent-ctl:${KATA_REF}-x86_64"
+	oras pull "${KATA_REG}/agent-ctl:${KATA_REF}-x86_64"
```
Sure, I can add a check in this branch but let it working when the user provides custom agent-ctl.

ldoktor commented on src/cloud-api-adaptor/podvm/hack/smoke_test.sh:
```diff
@@ -0,0 +1,179 @@
+		set +e
+		popd
+		[ -n "${SOCAT_PID}" ] && kill "${SOCAT_PID}"
+		sudo virsh destroy smoketest
```
I used `--transient` so destroy is sufficient.

ldoktor commented:
Changes:

* Added `set -euo pipefail`
* Added arch check when pulling kata-ctl
* Reworked the check to use only `Check` periodically (because CreateSandbox can only be called once so on partial success it could result in false negative), then on success follow-up with the original `CreateSandbox` and added also `DestroySandbox` just to expand the testing a bit.

ldoktor commented:
Changes:

* Added chmod +x to kata-ctl to ensure it's executable
* Added note about smoke-test to README file

ldoktor commented:
Changes:

* tabs vs. spaces
* removed `--skip-alias` from `which` (not supported by GH runners)
* moved `chmod +x` only to katactl from custom location

ldoktor commented:
Changes:

* Made the WORKDIR `a+rwx` to allow ourself using it for libvirt as well as for socat

ldoktor commented:
Changes:

* Added note about disabling tls for smoke testing

ldoktor commented on .github/workflows/podvm_smoketest.yaml:
```diff
@@ -57,110 +57,13 @@ jobs:
-          kata-agent-ctl connect \
-            --server-address "unix://${sock}" \
-            --cmd CreateSandbox
+          bash -e ../podvm/hack/smoke_test.sh -d build/podvm-fedora-amd64.qcow2
```
ack

ldoktor commented:
Changes:

* Removed "bash -e" call as suggested by @mkulke
* Rebased

ldoktor commented:
Changes:

* rebased
* added `require=off` for os variant
* defined SOCAT_PID
